sequencer/src/block2.rs

```diff
@@ -304,9 +314,52 @@ pub(super) type RangeProof =
 /// LargeRangeProof<KzgEval<Bls12_281>>
 /// ```
 /// but that's still pretty crufty.
-pub type NamespaceProof =
+pub type JellyfishNamespaceProof =
     LargeRangeProof<<UnivariateKzgPCS<Bls12_381> as PolynomialCommitmentScheme>::Evaluation>;
 
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct NamespaceProof<V>
+where
+    V: PayloadProver<JellyfishNamespaceProof>,
+{
+    ns_payload_flat: Vec<u8>,
+    ns_payload_range: Range<usize>,
+    ns_proof: JellyfishNamespaceProof,
+    ns_id: VmId,
+    vid_common: <V as VidScheme>::Common,
+}
+
+impl<V> NamespaceProof<V>
+where
+    V: PayloadProver<JellyfishNamespaceProof>,
+{
+    #[allow(dead_code)] // TODO temporary
+    pub fn verify(
+        &self,
+        vid: &V,
+        commit: &<V as VidScheme>::Commit,
+    ) -> Option<(Vec<Transaction>, VmId)> {
+        // verify self against args
+        vid.payload_verify(
+            Statement {
+                payload_subslice: &self.ns_payload_flat,
+                range: self.ns_payload_range.clone(),
+                commit,
+                common: &self.vid_common,
+            },
+            &self.ns_proof,
+        )
+        .ok()?
+        .ok()?;
+
+        // verification succeeded, return some data
+        Some((
+            parse_ns_payload(&self.ns_payload_flat, self.ns_id),
+            self.ns_id,
```

Comment:
Jellyfish doesn't know anything about namespaces, so `payload_verify` has no way to check `ns_id`.

Ugh if you want to authenticate `ns_id` then `verify` needs additional args: the namespace table (aka `metadata`) and the index for the namespace (so we know where to look in the namespace table to check `ns_id`).

I suppose you'll also want to authenticate the namespace table, too, as otherwise an attacker could make up any namespace table it needs to pass verification. I don't even know whether we currently have infra for that.

Reply:
In principle we could eliminate the need for these additional args to `verify` (that is, the namespace table and index) by storing them in the namespace proof like we do for everything else. But we'd still need some way to authenticate this data against a block commitment.

Reply:
Really all that is accomplished in the present PR is that we prove that *some* specific subslice of the payload has some specific bytes. We prove nothing about the ID of the namespace corresponding to that subslice, or even whether that subslice corresponds to a namespace at all.

This is going to require more work. We should meet to discuss.